Preserve the stack of the actual error.

papandreou opened the pull request:
Fixes #322.

I'm not totally happy with this, but maybe it's a special case that we can internalize fully.

papandreou commented:
@sunesimonsen, seen this one?

sunesimonsen commented:
Yes, I will look into it during the week.

sunesimonsen commented on lib/Unexpected.js:
```diff
@@ -488,6 +488,8 @@ Unexpected.prototype.fail = function (arg) {
             if (key === 'diff') {
                 error.createDiff = value;
+            } else if (key === 'stack') {
+                error._stack = value;
```
We should consider naming this property something else like `originalError` and store the store a reference to the error instance.

papandreou replied:
We could, but what would that accomplish, exactly? As far as I could tell we only need the `stack` trace of the original error (which includes the error message as well, in case we want that too).

sunesimonsen replied:
I'm fine with just having the stack - but I don't really like the half secret `_stack` field. I think we should make this a first class API using expect.fail.

sunesimonsen commented on lib/assertions.js:
```diff
@@ -595,11 +595,16 @@ module.exports = function (expect) {
+                }
+            }, function (e) {
+                e._stack = error && error.stack;
+                throw e;
```
An idea could be to use `expect.fail(e)` and always let fail with an error set the originalError.

papandreou replied:
That sounds like a really good idea.

papandreou replied:
Hmm, that already throws the original error, so the stack trace will presumably come out correct already: https://github.com/unexpectedjs/unexpected/blob/master/lib/Unexpected.js#L462-L464

papandreou replied:
Replacing

```js
                e._stack = error && error.stack;
                throw e;
```

with just

```js
                expect.fail(error);
```

seems to give exactly the same result. Would you prefer that?

sunesimonsen replied:
that looks cleaner - don't we need the `_stack` field then :grey_question:

sunesimonsen commented:
My comments are of cause up for discussion.

sunesimonsen commented:
We should be a bit careful if we only remember the stack, as a lot of test runners will use error message to figure out how to trim of the error message in the stack. So if you only set the stack and that contains a wrong error message it might get us into troubles. We should test the solution with Jasmine, Jasmine browser, Phantom and Mocha Browser.

papandreou commented:
> We should be a bit careful if we only remember the stack, as a lot of test runners will use error message to figure out how to trim of the error message in the stack

Ah, right, for some reason I didn't realize that `UnexpectedError.serializeMessage` left the error message at the beginning of the stack alone.

And when run through mocha that exact phenomenon you're describing occurs:

```
expected
function () {
  return expect.promise(function () {
    (function thisIsImportant() {
      throw new Error('argh');
    }());
  });
}
not to error
  returned promise rejected with: Error('argh')

  Error: argh
      at thisIsImportant (test/assertions/to-error.spec.js:114:31)
      at test/assertions/to-error.spec.js:115:22
      at lib/makePromise.js:60:38
      at Function.makePromise [as promise] (lib/makePromise.js:15:12)
      at test/assertions/to-error.spec.js:112:31
      at lib/assertions.js:615:24
      at lib/makePromise.js:60:38
      at Function.makePromise [as promise] (lib/makePromise.js:15:12)
      at Function.<anonymous> (lib/assertions.js:613:23)
      at executeExpect (lib/Unexpected.js:1121:50)
      at Unexpected.expect (lib/Unexpected.js:1129:22)
      at Context.<anonymous> (test/assertions/to-error.spec.js:111:20)
      set UNEXPECTED_FULL_TRACE=true to see the full stack trace
```

Where mocha would normally remove the `Error: argh` line.

papandreou commented:
@sunesimonsen Addressed all the comments except the idea of using `expect.fail(e, { originalError: originalError })` to fail instead of throw.

Turns out I couldn't really get that to work. It ended up resulting in failing tests with diffs like:

![errornestingdiff](https://cloud.githubusercontent.com/assets/373545/17754229/d1be4410-64d3-11e6-87b7-28502b29e412.png)

Here's what I tried: https://github.com/unexpectedjs/unexpected/commit/69082f4c1bc6996e521cabf3355d8db8de233945

sunesimonsen commented:
@papandreou awesome work. I'll just take the PR for a swing as the is very core, but other then that I think it is ready. Thanks.

sunesimonsen commented:
I tried to run the tests in phantom, but I get the following error:

```
[unexpected]$ make test-phantomjs 
(echo '/*!' && <LICENSE sed -e's/^/ * /' | sed -e's/\s+$//' && echo ' */' && ./node_modules/.bin/browserify -p bundle-collapser/plugin -e lib -s weknowhow.expect) > unexpected.js
phantomjs ./node_modules/mocha-phantomjs-core/mocha-phantomjs-core.js test/tests.html spec "`node -pe 'JSON.stringify({useColors:true,grep:process.env.grep})'`"
TypeError: 'undefined' is not an object (evaluating 'stderr.writeLine')

  ./node_modules/mocha-phantomjs-core/mocha-phantomjs-core.js:29
```

I don't think this is related, but it make it hard to know if the change i okay in Phantom.

sunesimonsen commented:
`make test-jasmine-browser` also fails with Failed: `setImmediate is not defined` and `ReferenceError: expectWithUnexpectedMagicPen is not defined` :cry:

sunesimonsen commented:
I still want to test this functionality in old IE browser, but I can only do that on Monday, as I left my work computer at the office.

papandreou commented:
>TypeError: 'undefined' is not an object (evaluating 'stderr.writeLine')
>
>  ./node_modules/mocha-phantomjs-core/mocha-phantomjs-core.js:29

It works for me, and also on Travis. [The line it's failing on](https://github.com/nathanboktae/mocha-phantomjs-core/blob/f645471008b145405e61a5234a1c9221b6d0feb8/mocha-phantomjs-core.js#L29) seems to indicate that it wanted to tell you that your version of Phantom.js is too old (<1.9.1), so could you try upgrading?

(@gustavnikolaj -- okay, so here we seem to have the first instance of the Phantom.js version problem you were worried about, duly noted)

papandreou commented:
Fixed the Jasmine setup on master (I hope) -- it was broken due to npm 3's deduping and me forgetting to introduce unexpected-magicpen in Jasmine.tpl.html.

Rebased feature/preserveStackOfThrownError on master after the fix landed.

sunesimonsen commented:
Cool I'll reevaluate.
On søn. 21. aug. 2016 at 00.54, Andreas Lind <notifications@github.com>
wrote:

> Fixed the Jasmine setup on master (I hope) -- it was broken due to npm 3's
> deduping and me forgetting to introduce unexpected-magicpen in
> Jasmine.tpl.html.
>
> Rebased feature/preserveStackOfThrownError on master after the fix landed.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/unexpectedjs/unexpected/pull/323#issuecomment-241228707>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFisuORb1Z7mjsmuMT1PNeueiYYiCYIks5qh4WPgaJpZM4JTVmf>
> .
>

sunesimonsen commented:
Now it all runs correctly on my old machine. Tomorrow I'll test it on old IE and then it should be ready to merge. Thanks for the effort.

sunesimonsen commented:
I have now tested this is IE and it works, so let's just merge it. Sorry for the long roundtrip.

sunesimonsen commented:
👍

papandreou commented:
Thanks :)

papandreou commented:
Released in 10.16.0
